src/adaptors/tcp_adaptor.c

```diff
@@ -1257,13 +1489,17 @@ static void qdr_tcp_delivery_update(void *context, qdr_delivery_t *dlv, uint64_t
     void* link_context = qdr_link_get_context(qdr_delivery_link(dlv));
     if (link_context) {
         qdr_tcp_connection_t* tc = (qdr_tcp_connection_t*) link_context;
-        qd_log(tcp_adaptor->log_source, QD_LOG_DEBUG, DLV_FMT" qdr_tcp_delivery_update: disp: %"PRIu64", settled: %s",
+        qd_log(tcp_adaptor->log_source, QD_LOG_DEBUG,
+               DLV_FMT" qdr_tcp_delivery_update: disp: %"PRIu64", settled: %s",
                DLV_ARGS(dlv), disp, settled ? "true" : "false");
 
         //
         // If one of the streaming deliveries is ever settled, the connection must be torn down.
```

Comment:
I think now that we have separate close calls for read and write sides, we should close for read if the instream is settled and close for write when the outstream is settled.

If we then settle and complete the instream on CLOSED_READ and settle the outstream on CLOSED_WRITE, we are then conveying more accurately what each tcp endpoint does to its peer. If the instream has not been started when we get CLOSED_READ on ingress, we can close the connection entirely (on egress we can just send an empty reply).

Reply:
I disagree. This code patch is trying to fix DISPATCH-1878 where a client sends data and closes the connection. The expectation is that the data gets sent over the wire to where it's going and the propagates the close after that. Any data will sent by the server will come back to the client. Then when the server closes that close will be propagated to the client.
A tcpListener READ_CLOSED event must be propagated to the peer tcpConnector as a write_close() control. Then the connector READ_CLOSE event becomes the listener write_close() control.
Short circuiting that control flow will lose data.

Reply:
That is exactly my point. The same is true for the other direction, a tcpListener CLOSED_WRITE should be propagated to the peer tcpConnector where it will call read_close(). The CLOSED_WRITE at the tcpConnector will be propagated to the tcpListener for it to call read_close(). Read and write streams are chained in both directions.

I argue that the comment above is actually short circuiting this (as it predates the API to handle read_close()/write_close() separately). Though it may work for the ncat case, that short circuiting seems less clean than a proper symmetric chaining.

Reply:
Maybe a solution would be not to accept an incoming client raw connection until after all of the TCP adaptor mechanisms have been set up. This would stall the initial connection setup but it would simplify propagating closed read/write in both directions when they happen.

Reply:
Readiness to read will always be a per-connection setup.

Reply:
Update on this. I was wrong. We don't actually explicitly settle deliveries except in the case where the connector cannot connect (i.e. disconnects before connecting), in which case the 'initial-delivery' is settled. Deliveries are implicitly settled if there is any failure. We convey the half clsoed state solely through completing the message (i.e. the EOS is the only trigger).

I've pushed a fix that handles the case where there is no server. Also tested router failures, including middle router in a three-way chain with tcp-listener and tcp-connector at opposite ends.

Reply:
https://github.com/ChugR/qpid-dispatch/pull/81